internal/domain/power-status.go

```diff
@@ -18,16 +19,730 @@
+			//error message reflect that.  Otherwise give a generic error.
+			pcomp := pcsmodel.PowerStatusComponent{XName: name}
+			htype := xnametypes.GetHMSType(name)
+			switch (htype) {
+				case xnametypes.Chassis:    fallthrough
+				case xnametypes.ChassisBMC: fallthrough
+				case xnametypes.NodeBMC:    fallthrough
+				case xnametypes.RouterBMC:  fallthrough
+				case xnametypes.Node:       fallthrough
+				case xnametypes.CabinetPDUOutlet:
+					pcomp.Error = "Component not found in component map."
+
+				case xnametypes.HMSTypeInvalid:
+					pcomp.Error = "Invalid component name."
+
+				default:
+					pcomp.Error = "Component can not have power state and managment state data"
+			}
+			rcomps.Status = append(rcomps.Status,pcomp)
+			continue
+		}
+		//Filter by pwrstate and mgmtstate
+		if (!psUndef) {
+			if (strings.ToLower(pwrStateFilter.String()) != strings.ToLower(mp.PowerState)) {
+				stateMatch = false
+			}
+		}
+		if (!msUndef) {
+			if (strings.ToLower(mgmtStateFilter.String()) != strings.ToLower(mp.ManagementState)) {
+				stateMatch = false
+			}
+		}
+
+		if (stateMatch) {
+			var cmp pcsmodel.PowerStatusComponent
+			cmp.SupportedPowerTransitions = make([]string,len(mp.SupportedPowerTransitions))
+			cmp.XName = name
+			cmp.PowerState = mp.PowerState
+			cmp.ManagementState = mp.ManagementState
+			copy(cmp.SupportedPowerTransitions,mp.SupportedPowerTransitions)
+			rcomps.Status = append(rcomps.Status,cmp)
+		}
+	}
+
+	robj = pcsmodel.BuildSuccessPassback(200,rcomps)
+	return robj
+}
+
+/////////////////////////////////////////////////////////////////////////////
+//                I N T E R N A L   F U N C T I O N S
+/////////////////////////////////////////////////////////////////////////////
+
+// Update the cached HW state map using HSM.  This will add new components
+// to the map but not the actual HW states.
+
+func updateComponentMap() error {
+	fname := "updateComponentMap()"
+
+	//Get all components in HSM
+
+	compMap,err := (*hsmHandle).FillHSMData([]string{"all"})
+	if (err != nil) {
+		return fmt.Errorf("Error fetching HSM data: %v",err)
+	}
+
+	if (len(compMap) == 0) {
+		return fmt.Errorf("HSM returned empty list of components!")
+	}
+
+	//TODO: Not sure if this is kosher... we'll remove any entry from
+	//our in-memory component map if it is not returned by HSM.  That way
+	//the in-memory map is "in sync" with HSM.   If for whatever reason
+	//components are removed from HSM, we'll remove them from our internal
+	//map too, so we don't keep trying to contact BMCs that may not exist.
+
+	for k,_ := range(hwStateMap) {
+		_,ok := compMap[k]
+		if (!ok) {
+			glogger.Infof("Removing '%s' from local map (no longer in HSM component list).", k)
+			delete(hwStateMap,k)
+			(*kvStore).DeletePowerStatus(k)
+		}
+	}
+
+	//Filter on all pertinent component types and add to the component map.
+
+	for _,v := range(compMap) {
+		switch (xnametypes.HMSType(v.BaseData.Type)) {
```

Comment:
Don't forget to add the missing types here and the other locations:
ComputeModule
RouterModule
CabinetPDUPowerConnector

Reply:
Why?  We discussed these and decided that they don't have the support in HSM, so we
don't want to infer anything for these types.

Reply:
Done.